pkg/core/writer/tracetracker/tracker.go

```diff
@@ -20,44 +20,88 @@ const spanCorrelationMetricName = "sf.int.service.heartbeat"
 // they are not seen for a certain amount of time.
 type ActiveServiceTracker struct {
 	sync.Mutex
-	// How long to keep sending metrics for a particular service name after it
-	// is last seen
-	timeout time.Duration
-	// A linked list of serviceNames sorted by time last seen
-	serviceNameByTime *list.List
-	// Which service names are active currently.  The value is an entry in the
-	// serviceNameByTime linked list so that it can be quickly accessed and
-	// moved to the back of the list.
-	serviceNamesActive map[string]*list.Element
+
+	// hostIDDims is the map of key/values discovered by the agent that identify the host
+	hostIDDims map[string]string
+
+	// sendTraceHostCorrelationMetrics turns metric emission on and off
+	sendTraceHostCorrelationMetrics bool
+
+	// hostServiceCache is a cache of services associated with the host
+	hostServiceCache *TimeoutCache
+
+	// hostEnvironmentCache is a cache of environments associated with the host
+	hostEnvironmentCache *TimeoutCache
+
+	// tenantServiceCache is the cache for services related to containers/pods
+	tenantServiceCache *TimeoutCache
+
+	// tenantEnvrionmentCache is the cache for environments related to containers/pods
+	tenantEnvironmentCache *TimeoutCache
+
 	// Datapoints don't change over time for a given service, so make them once
 	// and stick them in here.
 	dpCache map[string]*datapoint.Datapoint
+
 	// A callback that gets called with the correlation datapoint when a
 	// service is first seen
 	newServiceCallback func(dp *datapoint.Datapoint)
 	timeNow            func() time.Time
 
+	// correlationClient is the client used for updating infrastructure correlation properties
+	correlationClient correlations.CorrelationClient
+
 	// Internal metrics
-	activeServiceCount int64
-	purgedServiceCount int64
-	spansProcessed     int64
+	spansProcessed int64
 }
 
-type serviceStatus struct {
-	LastSeen    time.Time
-	ServiceName string
+// LoadCorrelations asynchronously retrieves all known correlations from the backend
+// for all known hostIDDims.  This allows the agent to timeout and manage correlation
+// deletions on restart.
+func (a *ActiveServiceTracker) LoadCorrelations() {
+	// asynchronously fetch all services and environments for each hostIDDim at startup
+	for dimName, dimValue := range a.hostIDDims {
+		dimName := dimName
+		dimValue := dimValue
+		a.correlationClient.Get(dimName, dimValue, func(correlations map[string][]string, err error) {
+			// if there was an error unmarshalling the response,
+			// then there's not much we can do.  Just move on.
+			if err != nil {
+				return
```

Comment:
I would make sure this is logged either here or higher up.

Reply:
Also if this is just an error about unmarshalling, what happens if the server can't be reached at all?

Reply:
If the server can't be reached the error is handled by the client.  This call back will just get an empty payload and the error.

Reply:
I've added logging for failures to unmarshal the payload